proposals/NNNN-add-preferred-FPS.md

```diff
@@ -0,0 +1,128 @@
+
+## Detailed design
+
+### RPC change (Mobile_API and HMI_API)
+
+Add preferredFPS to VideoStreamingCapability struct in both APIs.
+
+```xml
+    <struct name="VideoStreamingCapability" since="4.5">
+        <description>Contains information about this system's video streaming capabilities.</description>
+        ...
+        <!-- new param -->
+        <param name="preferredFPS" type="Integer" minvalue="0" maxvalue="2147483647" mandatory="false">
+            <description>Preferred frame rate per second. Mobile application should take this value into account for capturing and encoding video frame.</description>
+        </param>
+    </struct>
+```
+
+Like other parameters in VideoStreamingCapability, preferredFPS should be returned in GetSystemCapability RPC response.
+
+### iOS mobile proxy changes
+
+Extend SDLVideoStreamingCapability class to have preferredFPS property:
+
+SDLVideoStreamingCapability.h:
+
+```objc
+/**
+ *  Preferred frame rate (assuming the case where resolution is the same as preferredResolution)
+ *  Optional
+ */
+@property (nullable, strong, nonatomic) NSNumber<SDLInt> *preferredFPS;
+
+```
+
+SDLVideoStreamingCapability.m:
+```objc
+- (void)setPreferredFPS:(nullable NSNumber<SDLInt> *)preferredFPS {
+    [store sdl_setObject:preferredFPS forName:SDLRPCParameterNamePreferredFPS];
+}
+
+- (nullable NSNumber<SDLInt> *)preferredFPS {
+    return [store sdl_objectForName:SDLRPCParameterNamePreferredFPS ofClass:NSNumber.class error:nil];
+}
+```
+
+And then, modify SDLStreamingVideoLifecycleManager#didEnterStateVideoStreamStarting, so that preferredFPS value would be taken into account:
+```objc
+static NSUInteger const defaultFrameRate = 15;
+
+- (void)didEnterStateVideoStreamStarting {
+    ...
+    if (capability != nil) {
+        NSUInteger  preferredFPS = (capability.preferredFPS.integerValue == 0) ? defaultFrameRate : capability.preferredFPS.integerValue;
+        weakSelf.videoEncoderSettings[(__bridge NSString *)kVTCompressionPropertyKey_ExpectedFrameRate] = @(preferredFPS);
+        ...
+    } else {
+        ...
+        weakSelf.videoEncoderSettings[(__bridge NSString *)kVTCompressionPropertyKey_ExpectedFrameRate] = @(defaultFrameRate);
+    }
+
+}
+```
+Note that videoEncoderSettings property must be NSMutableDictionary as it now would be mutable.
+
+### iOS mobile proxy consideration
+
+A frame rate of iOS SDL application highly depends on how the application captures the video data.
+
+If an application uses combination of SDLCarWindow and SDLStreamingVideoLifecycleManager, above change should take account of
+1) videoEncoderSettings used in SDLStreamingVideoLifecycleManager,
+2) capture rate of CADisplayLink#preferredFramePerSecond (available in iOS10 or higher)
+
+So that should suffice the needs.
+
+If an application uses AVCaptureDevice directly, and does not rely on SDLStreamingVideoLifecycleManager at all, the frame rate must fit into AVFrameRateRange in ACCaptureDevice.format.

```

Comment:
```suggestion
If an application uses `AVCaptureDevice` directly, and does not rely on `SDLStreamingVideoLifecycleManager` at all, the frame rate must fit into `AVFrameRateRange` in ACCaptureDevice.format.
```

Reply:
Fixed typo (ACCaptureDevice to AVCaptureDevice) as well.